pyface/data_view/i_data_view_widget.py

```diff
@@ -24,6 +27,108 @@
 logger.level = logging.DEBUG
 
 
+class _Row(BaseTuple):
+    """ Trait type for validating a row index, used internally by
+    an DataViewWidget.
+
+    This trait type relies on the following attributes being defined on the
+    object:
+    - data_model
+    - selection_type
+    """
+
+    def validate(self, object, name, value):
+        row = super(_Row, self).validate(object, name, value)
+        if not object.data_model.is_row_valid(row):
+            raise TraitError("Invalid row index {!r}".format(row))
+
+        if object.selection_type == 'item':
+            can_have_children = object.data_model.can_have_children(row)
+            have_rows = object.data_model.get_row_count(row) > 0
+            if can_have_children and have_rows:
+                raise TraitError(
+                    "Row values must have no children when selection_type "
+                    "is 'item', got {!r}".format(row)
+                )
```

Comment:
I added the validation for 'item' following the rest of the validation logic. But I have not added the code in the toolkit specific widget to actually avoid setting an invalid selection. e.g. if the user tries to click on row and column, the toolkit specific widget at the moment does not filter the event. Then we will get a trait error, and in Qt case, a hard crash. That filtering will need to be done if we include this validation.

Reply:
Or I might have misunderstood what "item" means. The above code tries to exclude row and column headers, but maybe "item" is meant to include them? Headers do tend to represent different data though. 🤫